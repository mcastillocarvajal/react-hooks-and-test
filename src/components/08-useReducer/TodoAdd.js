import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handleAddTodo }) => {

    const [ { description }, handleInputChange, reset] = useForm({
        description: '',
    })

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 0 ){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddTodo( newTodo )
        reset();
    }

    return (
        <>
            <h4> Add todo </h4>
            <hr />

            <form onSubmit={ handleSubmit }>
                <input
                    className='form-control'
                    type='text'
                    name='description'
                    placeholder='Learn...'
                    autoComplete='off'
                    value={ description }
                    onChange={ handleInputChange }
                />

                <button 
                    className='btn btn-outline-primary mt-1 btn-block'
                    type='submit'>
                    Add
                </button>
            </form>
        </>
    )
}
