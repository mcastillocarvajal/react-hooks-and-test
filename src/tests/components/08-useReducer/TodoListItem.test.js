import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Test for <TodoListItem />', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow( 
        <TodoListItem
            todo={ demoTodos[0] }
            index={ 0 }
            handleDelete={ handleDelete }
            handleToggle={ handleToggle }
        />
    )

    test('should show component correctly', () => {
        
        expect( wrapper ).toMatchSnapshot();
    })
    
    test('should call handleDelete', () => {
        
        wrapper.find('button').simulate('click');
        expect( handleDelete ).toHaveBeenCalledWith( demoTodos[0].id )
    })

    test('should call handleToggle', () => {

        wrapper.find('p').simulate('click');
        expect( handleToggle ).toHaveBeenCalledWith( demoTodos[0].id )
    })

    test('should show the text correctly', () => {

        const p = wrapper.find('p');
        expect( p.text().trim() ).toBe( `${ 0 + 1 }. ${ demoTodos[0].desc }` )
        
    })

    test('should have classname complete if todo.done = true', () => {

        const todo = demoTodos[0];
        todo.done = true;

        const wrapper = shallow( 
            <TodoListItem
                todo={ todo }
            />
        )

        expect( wrapper.find('p').hasClass('complete') ).toBe(true);
    })
    
})
