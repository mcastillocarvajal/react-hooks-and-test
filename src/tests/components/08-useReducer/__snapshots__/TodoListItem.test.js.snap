// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for <TodoListItem /> should show component correctly 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . 
    Learn React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
